inject/titus-nsenter.c

```diff
@@ -0,0 +1,260 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <linux/limits.h>
+#include <errno.h>
+#include <string.h>
+
+/* prctl  / waitpid */
+#include <sys/prctl.h>
+#include <signal.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+/* Open */
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+/* Unshare */
+#include <sched.h>
+#include <unistd.h>
+
+#include "shared.h"
+
+struct namespace {
+	int nstype;
+	char *name;
+};
+
+static struct namespace namespaces[] = {
+	{
+		.nstype	= CLONE_NEWNS,
+		.name	= "mnt",
+	},
+	{
+		.nstype	= CLONE_NEWCGROUP,
+		.name 	= "cgroup",
+	},
+	{
+		.nstype	= CLONE_NEWIPC,
+		.name	= "ipc",
+	},
+	{
+		.nstype	= CLONE_NEWNET,
+		.name	= "net",
+	},
+	{
+		.nstype	= CLONE_NEWUSER,
+		.name	= "user",
+	},
+	{
+		.nstype	= CLONE_NEWUTS,
+		.name	= "uts",
+	},
+	/* Order here is intentional. We do pid NS last. */
```

Comment:
Why do we do pid last?

Reply:
It opens up an exploit path if we do PID before any of the other ones. Specifically, the attack vector is that the container shows up in the PID namespace (/proc/...) in the container before it's changed over all its namespaces to container namespaces. 

If someone breaks in using some new vector that we don't know of while this change is happening, they can traverse back up to the host OS's namespaces.